### PR TITLE
Parse autolinks for urls and emails.

### LINKED ISSUE #9564: Markdown fails to parse Autolinks
It seems that Markdown.parse in 0.4 doesn't parse autolinks (http://spec.commonmark.org/0.14/#autolink).

``` julia
julia> Markdown.parse("<http://github.com>")
  <http://github.com>
```

As a related question, should I report Markdown bugs here or to `Markdown.jl`? This is the third Markdown bug I come across today and reported the two others to `Markdown.jl`, but they also exist in `base.Markdown`.

### PR DESCRIPTION
**Pull request by hayd**
fixes #9564 

cc @one-more-minute

### CODE REVIEW

**Comment by hayd** [comment]
I forgot that this fails due to `@` being escaped. It's probably safe to remove that from the chars we escape (although perhaps the mailto link still works?? - I'm not sure).

Unfortunately/interestingly I don't think commonmark implementation actually follows it's spec with regards to scheme-sensitivity. I don't think we're following the spec on what/when to escape chars... similarly the commonmark implementation is also not following the spec there either!

**Comment by one-more-minute** [comment]
Sorry for taking a ridiculously long time to look at this. Looks good anyway. I don't mind removing `@` from the escapes to fix this. The only other thing I'd say to to read until `>`, `[space]` or `\n` – for the sake of efficiency it makes sense to stop reading the stream as early as possible.

**Comment by tkelman** [comment]
Has this been done in Documenter?

**Comment by MichaelHatherly** [comment]
> Has this been done in Documenter?

Needs to be part of `Base.Markdown`, so would be nice to get this rebased and finished up (if @hayd has spare time to do it?)

**Comment by hayd** [comment]
The rebase was over footnotes, seemed to be trivial. Pushed but haven't tested...

I'm yet to action on:

> The only other thing I'd say to to read until >, [space] or \n – for the sake of efficiency it makes sense to stop reading the stream as early as possible.

What's a clean way to do that?

**Comment by MichaelHatherly** [comment]
> What's a clean way to do that?

Does `Markdown.startswith` (with a `Regex` as the second argument) do what you need?

File: base/markdown/Common/inline.jl [hunk 1]
```diff
@@ -106,6 +106,42 @@ function footnote_link(stream::IO, md::MD)
     end
 end
 
+@trigger '<' ->
+function autolink(stream::IO, md::MD)
+    withstream(stream) do
+        startswith(stream, '<') || return
+        url = readuntil(stream, '>')
```
Comment (mpastell) [inline base/markdown/Common/inline.jl]: Try `url = readuntil(stream, '>', match = ' ') `

**Review by mpastell — COMMENTED** [review]
I added a suggestion for reading. It would nice to have this merged for 0.6

**Comment by StefanKarpinski** [comment]
Bump, would love to have this. Anyone know what's busted on CI here?

**Comment by tkelman** [comment]
failed doc build on travis and failed tests on appveyor, due to the `_is_link` function somehow getting a `nothing` input?

File: base/markdown/Common/inline.jl (hunk 1 above)
Comment (hayd) [inline base/markdown/Common/inline.jl]: Are we hitting this definition already? https://github.com/JuliaLang/julia/blob/550f5d33cad1396e6e9a84843bddc8cc9b4702ed/base/markdown/parse/util.jl#L139-L159

I guess I haven't fully grokked what match/count do there, but it looked to me like it was suggesting you'd use `<` as the match character i.e. matching braces.

**Comment by hayd** [comment]
> Returns nothing and resets the stream if delim is not found.

So that's a bug, it should be returning false if url is nothing. Perhaps implementation changed somewhere down the line, at any rate will push a fix this evening.

File: base/markdown/Common/inline.jl (hunk 1 above)
Comment (mpastell) [inline base/markdown/Common/inline.jl]: Sorry, I misread it. I think you're right and need to do:

`text = readuntil(stream, '>', match = '<')`

I originally read it too quickly and thought setting it to `' '` to would stop reading at whitespace.

Comment (mpastell) [inline base/markdown/Common/inline.jl]: I think `readuntil` returns `nothing` if the input doesn't have '>'. Try with e.g. `Markdown.parse("<link")`. Try to add  `url ≡ nothing && return`

**Comment by hayd** [comment]
This is now green.

File: base/markdown/Common/inline.jl [hunk 2]
```diff
@@ -106,6 +106,43 @@ function footnote_link(stream::IO, md::MD)
...
+        url ≡ nothing && return
+        _is_link(url) && return Link(url, url)
+        _is_mailto(url) && return Link(url, url)
+        return
+    end
+end
+
+# This list is take from the commonmark spec
```
Comment (tkelman) [inline base/markdown/Common/inline.jl]: is taken

File: base/markdown/Common/inline.jl [hunk 3]
```diff
@@ -106,6 +106,43 @@ function footnote_link(stream::IO, md::MD)
...
+adiumxtra afp afs aim apt,attachment aw beshare bitcoin bolo callto chrome,chrome-extension com-eventbrite-attendee content cvs,dlna-playsingle dlna-playcontainer dtn dvb ed2k facetime feed finger fish gg git gizmoproject gtalk hcp icon ipn irc irc6 ircs itms jar jms keyparc lastfm ldaps magnet maps market,message mms ms-help msnim mumble mvn notes oid palm paparazzi platform proxy psyc query res resource rmi rsync rtmp secondlife sftp sgn skype smb soldat spotify ssh steam svn teamspeak
+things udp unreal ut2004 ventrilo view-source webcal wtai wyciwyg xfire xri ymsgr"))
+
+function _is_link(s::AbstractString)
+    '<' in s && return false
+
+    m = match(r"^(.*)://(\S+?)(:\S*)?$", s)
+    m == nothing && return false
```
Comment (tkelman) [inline base/markdown/Common/inline.jl]: `=== nothing`

**Comment by tkelman** [comment]
Remind me to merge this in a day or two if we haven't gotten any other feedback.

**Comment by KristofferC** [comment]
You should merge this in a day or two.

**Review by MichaelHatherly — APPROVED** [review]

**Comment by tkelman** [comment]
I meant more like now.

**Pull request merged**